clerk.go

```diff
@@ -0,0 +1,360 @@
+	SetBackend(b)
+	return b
+}
+
+// SetBackend sets the Backend that will be used to make requests
+// to the Clerk API.
+// Use this method if you need to override the default Backend
+// configuration.
+func SetBackend(b Backend) {
+	backend.mu.Lock()
+	defer backend.mu.Unlock()
+	backend.Backend = b
+}
+
+type defaultBackend struct {
+	HTTPClient *http.Client
+	URL        string
+}
+
+// Call sends requests to the Clerk API and handles the responses.
+func (b *defaultBackend) Call(ctx context.Context, apiReq *APIRequest, setter ResponseReader) error {
+	req, err := b.newRequest(ctx, apiReq)
+	if err != nil {
+		return err
+	}
+
+	return b.do(req, apiReq.Params, setter)
+}
+
+func (b *defaultBackend) newRequest(ctx context.Context, apiReq *APIRequest) (*http.Request, error) {
+	path := b.URL + apiReq.Path
+	req, err := http.NewRequest(apiReq.Method, path, nil)
+	if err != nil {
+		return nil, err
+	}
+	req.Header.Add("Authorization", fmt.Sprintf("Bearer %s", SecretKey))
+	req.Header.Add("Content-Type", "application/json")
+	req.Header.Add("User-Agent", fmt.Sprintf("Clerk/%s SDK-Go/%s", clerkVersion, sdkVersion))
+	req.Header.Add("X-Clerk-SDK", fmt.Sprintf("go/%s", sdkVersion))
+	req = req.WithContext(ctx)
+
+	return req, nil
+}
+
+func (b *defaultBackend) do(req *http.Request, params Queryable, setter ResponseReader) error {
+	err := setRequestBody(req, params)
+	if err != nil {
+		return err
+	}
+
+	resp, err := b.HTTPClient.Do(req)
+	if err != nil {
+		return err
+	}
+	resBody, err := io.ReadAll(resp.Body)
+	defer resp.Body.Close()
+	if err != nil {
+		return err
+	}
+
+	apiResponse := NewAPIResponse(resp, resBody)
+	// Looks like something went wrong. Handle the error.
+	if !apiResponse.Success() {
+		return handleError(apiResponse, resBody)
+	}
+
+	setter.Read(apiResponse)
+	err = json.Unmarshal(resBody, setter)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Sets the params in either the request body, or the querystring
+// for GET requests.
+func setRequestBody(req *http.Request, params Queryable) error {
+	// GET requests don't have a body, but we will pass the params
+	// in the query string.
+	if req.Method == http.MethodGet {
+		q := req.URL.Query()
+		params.Add(q)
+		req.URL.RawQuery = q.Encode()
+		return nil
+	}
+
+	body, err := json.Marshal(params)
+	if err != nil {
+		return err
+	}
+	req.Body = io.NopCloser(bytes.NewReader(body))
+	req.GetBody = func() (io.ReadCloser, error) {
+		return io.NopCloser(bytes.NewReader(body)), nil
+	}
+
+	return nil
+}
+
+// Error response handling
+func handleError(resp *APIResponse, body []byte) error {
+	apiError := &APIErrorResponse{
+		HTTPStatusCode: resp.StatusCode,
+	}
+	apiError.Read(resp)
+	err := json.Unmarshal(body, apiError)
+	if err != nil || apiError.Errors == nil {
+		// This is probably not an expected API error.
+		// Return the raw server response.
+		return errors.New(string(body))
+	}
+	return apiError
+}
+
+// The active Backend
+var backend api
```

Comment:
❓ I'm not sure I understand the use of these globals, along with the `Get/SetBackend` functions. I would maybe expect that a `NewAPIClient` method (if I understand the parity with the `Backend` interface here), returns a fully isolated client instance. Any chance we need to simplify some aspects here?

Reply:
It depends. 

Having a "singleton" global API client allows for two things:

1. Simple API for operations
```go
clerk.SecretKey = "sk_live_123"
dmn, err := domain.Create(ctx, &DomainParams{})
```

2. Ability to stub out the client for tests
```go
clerk.SetBackend(clerk.NewBackend(&clerk.BackendConfig{
  HTTPClient: myCustomClient,
}))

domain.Create(ctx, &clerk.DomainParams{})
```

I'm facing the same dilemma as you. Is the global singleton-like backend worth the trouble? Will it actually prove flexible or not? I toyed around with different approaches, but DX-wise, I believe nothing beats the API described above.

How would the alternative look like?
```go
domain.NewClient(&clerk.ClientConfig).Create(ctx, &clerk.DomainParams{})
```

Or I imagine in your main.go file
```go
package main

type Server struct {
  domainClient *domain.Client
  organizationClient *organization.Client
}

func main() {
  // Instantiate your clients and your server struct to ListenAndServe
  clerkConfig := &clerk.BackendConfig{
    HTTPClient: myCustomClient,
  }
  srv := &Server{
    domainClient: domain.NewClient(clerkConfig),
    organizationClient: organization.NewClient(clerkConfig),
  }
}

// and then in your handler
func (srv *Server) CreateDomains(w, r) {
  dmn, err := srv.domainClient.Create(ctx, &clerk.DomainParams{})
}
```

Comparing this solution, to the `dmn, err := domain.Create()` example above, there's a clear DX benefit in my eyes.

What do you think?

Reply:
> Ability to stub out the client for tests

The use of a global variable here seems to be breaking test isolation, if they are running in parallel. So if I understand correctly we may not allowing dependency injection.

I think you can avoid the use of globals with defining an explicit setter for a default backend and provide constructors for dependency injection. I imagine the API being something like:


```go
package clerk

type Backend struct {
  httpClient *http.Client
}

var defaultBackend = &Backend{}
func SetDefaultBackend(cfg *BackendConfig) {
  // Resembles `SetBackend` now but mutates an existing struct
  defaultBackend.httpClient = cfg.HTTPClient
}

func DefaultBackend() *Backend {
  return defaultBackend
}

type DomainsClient struct {
  backend *Backend
}

func NewDomainsClient(backend *Backend) {}

var Domains = struct {
  backend *Backend
}{backend: defaultBackend}

func (d *Domains) Create(ctx context.Context, params *CreateDomainParams) (Domain, error) {}
```

```go
package main

func main() {
  // or this can happen on the user's package-level init()
  clerk.SetDefaultBackend(...)
}

func CreateDomains(ctx context.Context, name string) (..., error) {
  domain, err := clerk.Domains.Create(ctx, &clerk.DomainParams{Name: name})
}
```

Just an idea, I'm not sure what the exact end-user API needs to be.

Reply:
Resolved the discussion offline. We'll go ahead with the current design until there's more strong evidence towards change.